_posts/2021-04-26-4.0.0-release.md

```diff
@@ -0,0 +1,161 @@
+type instead of erroring when the types are not equal.
+
+ARROW-11744 - [C++] Add xsimd dependency
+
+Compute functions `quantile` (ARROW-10831) and `power` (ARROW-11070) have been
+added for numeric data.
+
+Compute functions for string processing have been added for:
+
+* Trim characters (ARROW-9128).
+* Extract substrings captured by a regex pattern (`extract_regex`, ARROW-10195).
+* Compute UTF8 string lengths (`utf8_length`, ARROW-11693).
+* Match strings against regex pattern (`match_substring_regex`, ARROW-12134).
+* Replace non-overlapping substrings that match (regex) pattern with
+  replacement (`replace_substring` and `replace_substring_regex`,
+  ARROW-10306).
+
+### Dataset
+
+TODO, possible issues to mention:
+
+ARROW-10438 - [C++][Dataset] Partitioning::Format on nulls
+ARROW-10846 - [C++] Add async filesystem operations
+ARROW-11174 - [C++][Dataset] Make Expressions available for projection
+ARROW-11601 - [C++][Dataset] Expose pre-buffering in ParquetFileFormatReaderOptions
+
+
+## C# notes
+
+## Go notes
+
+The go implementation now supports IPC buffer compression
+
+## Java notes
+Java now supports IPC buffer compression (ZSTD is recommended as the current performance of LZ4 is quite slow).
+
+## JavaScript notes
+* The Arrow JS module is now tree-shakeable. 
+* Iterating over Tables or Vectors is ~2X faster. [Demo](https://observablehq.com/@domoritz/arrow-js-3-vs-4-iterator) 
+* The default bundles use modern JS.
+## Python notes
+- Limited support for writing out CSV files (only types that have cast implementation to String) is now available.
+- Writing parquet list types now has the option of enabling the canonical group naming according to the Parquet specification.
+- The ORC Writer is now available.
+
+Creating a dataset with `pyarrow.dataset.write_dataset` is now possible from a
+Python iterator of record batches (ARROW-10882).
+The Dataset interface can now use custom projections using expressions when
+scanning (ARROW-11750). The expressions gained basic support for arithmetic
+operations (e.g. `ds.field('a') / ds.field('b')`) (ARROW-12058). See
+the [Dataset docs][7] for more details.
+
+See the C++ notes above for additional details.
+
+## R notes
+
+The `dplyr` interface to Arrow data gained many new features in this release, including support for `mutate()`, `relocate()`, and more. You can also call in `filter()` or `mutate()` over 100 functions supported by the Arrow C++ library, and many string functions are available both by their base R (`grepl()`, `gsub()`, etc.) and `stringr` (`str_detect()`, `str_replace()`) spellings.
+
+Datasets can now read compressed CSVs automatically, and you can also open a dataset that is based on a single file, enabling you to use `write_dataset()` to partition a very large file without having to read the whole file into memory.
+
+For more on what’s in the 4.0.0 R package, see the [R changelog][4].
+
+## Ruby and C GLib notes
+
+### Ruby
+
+### C GLib
+
+## Rust notes
+
```

Comment:
```suggestion
This release of Arrow continues to add new features and performance improvements. Much of our time this release was spent hammering out the necessary details so we can release the Rust versions to cargo at a more regular rate. In addition, we welcomed the Ballista distributed compute project officially to the fold. 

### Arrow

- Improved LargeUtf8 support

- Improved null handling in AND/OR kernels

- Added JSON writer support (ARROW-11310)

- JSON reader improvements

- - LargeUTF8
  - Improved schema inference for nested list and struct types

- Various performance improvements

- IPC writer no longer calls finish() implicitly on drop

- Compute kernels

- - Support for optional `limit` in sort kernel 
  - Divide by a single scalar
  - Support for casting to timestamps
  - Cast: Improved support between casting List, LargeList, Int32, Int64, Date64
  - Kernel to combine two arrays based on boolean mask
  - Pow kernel

- `new_null_array` for creating Arrays full of nulls. 

### Parquet

- Added support for filtering row groups (used by DataFusion to implement filter push-down)
- Added support for Parquet v 2.0 logical types 

### DataFusion

New Features

- SQL Support

- - CTEs
  - UNION
  - HAVING
  - EXTRACT
  - SHOW TABLES
  - SHOW COLUMNS
  - INTERVAL
  - SQL Information schema
  - Support GROUP BY for more data types, including dictionary columns, boolean, Date32

- Extensibility API

- - Catalogs and schemas support
  - Table deregistration
  - Better support for multiple optimizers 
  - User defined functions can now provide specialized implementations for scalar values

- Physical Plans

- - Hash Repartitioning

- SQL Metrics

- Additional Postgres compatible function library:

- - Length functions
  - Pad/trim functions
  - Concat functions
  - Ascii/Unicode functions
  - Regex

- Proper identifier case identification (e.g. “Foo” vs Foo vs foo)

- Upgraded to Tokio 1.x

  

Performance Improvements:

- LIMIT pushdown
- Constant folding
- Partitioned hash join
- Create hashes vectorized in hash join
- Improve parallelism using repartitioning pass
- Improved hash aggregate performance with large number of grouping values
- Predicate pushdown support for table scans
- Predicate push-down to parquet enables DataFusion to quickly eliminate entire parquet row-groups based on query filter expressions and parquet row group min/max statistics



API Changes

- DataFrame methods now take `Vec<Expr>` rather than `&[Expr]`
- TableProvider now consistently uses `Arc<TableProvider>` rather than `Box<TableProvider>`

### Ballista

- Ballista was donated shortly before the Arrow 4.0.0 release and there is no new release of Ballista as part of Arrow 4.0.0
```